### DISCUSSION TITLE
Issues with line offsetting and multiple line stacking

### DISCUSSION THREAD

**Discussion by pdrhlik** [systemed/tilemaker#712] [Q&A]
Hi,

I'm playing around with placing coloured hiking routes next to paths/roads. It's easy to place them on top of the paths but a lot of information is hidden because they overlap. I tried using the `line-offset` property in Maputnik which generally does the job. Unfortunately, it happens quite often that there are two ways in the opposite direction and the `line-offset` property reacts to that and offsets ways going forward to the right and ways going backward to the left. This creates too many undesired lines. Take a look at the following picture:

![Screenshot 2024-04-26 at 13 02 10](https://github.com/systemed/tilemaker/assets/14300398/b95d0310-3a9e-4243-990c-773f5d994712)

I would only like to see one of the red lines here. I don't care about the direction of ways and I would like to offset all of them to one side. If it was possible to determine the direction of a way/relation in `tilemaker`, then I could adjust the `line-offset` property accordingly. Is that possible?

I would also like to offset more coloured lines next to each other if they are present. This is the roughly the desired result that I would like to achieve. A path and multiple coloured lines right next to it. The number of these coloured paths can be anywhere from zero to four or five.

![Screenshot 2024-04-26 at 13 10 37](https://github.com/systemed/tilemaker/assets/14300398/998c2372-15c4-4aea-9efc-b209b05df520)

My main question probably is if this is something solvable on `tilemaker` level or if it's only probably a style issue.

Thanks for any input!

Cheers,
Patrik

**Comment by SomeoneElseOSM** [comment]
(no idea about the offset question, above my pay grade, but...)

on the "information lost because data overlaps" side of things, can you perhaps try experimenting with "casing" and "fill" (on the road) and very widely spaced dots (on the hiking route) so that you can see the road underneath?  https://map.atownsend.org.uk/maps/map/map.html#18/53.89625/-2.22001 is a Carto CSS style, but broadly the same controls are available to vector styles too.  I've not tried a vector map showing hiking routes yet, but 
<img width="101" alt="Screenshot 2024-04-26 132528" src="https://github.com/systemed/tilemaker/assets/5311740/40f995e7-1242-4ed4-a61a-da950233eff7">
(vector) is based on
<img width="98" alt="Screenshot 2024-04-26 132645" src="https://github.com/systemed/tilemaker/assets/5311740/7efb682f-c7b4-43b2-977d-30ed5ced2e44">
(raster), and mostly the same things are used to define it, with the addition of "stops".

**Reply by pdrhlik** [reply]
The dot approach is valid, I've seen in a few different styles already. The thing is that I don't subjectively like it. I'm used to a certain style that everybody is used to in Czech republic: https://en.mapy.cz/s/cahekerura

![Screenshot 2024-04-26 at 14 37 44](https://github.com/systemed/tilemaker/assets/14300398/846df7f1-ac2f-4593-8ba9-67512939b0a5)

Which is why I would like to achieve something similar. I can imagine that "casing" might be a way to get around it. The thing is that the lines would be symmetric and thus duplicated.

The "offsetting" technique sounds and looks exactly like what I'm trying to achieve but it just doesn't work all the way.

Thanks for the suggestions anyway!

**Comment by systemed** [comment]
tilemaker doesn't have any capability to offset a line itself (that sort of heavy geometry lifting is really something you need PostGIS for) or anything to determine the direction of a way.

I don't think the latter would be impossible to achieve - we already have Lua functions to calculate the area of a way, for example. But I'm not sure how exactly we'd do it. (Maybe a Lua accessor for the start lat/lon and end lat/lon of the line, and you could then calculate the bearing yourself in Lua? Would that work?)

I am a bit bemused as to why it's generating two lines though. Is this two different hiking relations or something?

**Reply by pdrhlik** [reply]
That should be enough! If I would know the start and end point of the line, it should be easy to determine the direction in most situations. I'm not sure I would be able to write that functionality myself though. I guess I could try and learn from the `Area` functionality. But that would probably sufficiently solve my issue. I would then just "hard code" line offsets per color (red is always next to the path, then blue, etc.). That would create gaps but I could live with that.

Exactly, there is more than one relation that overlap here. It happens quite often. As I said before, the offset then reacts based on the direction of the line:

![Screenshot 2024-04-26 at 15 00 11](https://github.com/systemed/tilemaker/assets/14300398/e1f99de3-c851-4522-bb29-c4b4697e5479)
